foundationdb/examples/hello.rs

```diff
@@ -0,0 +1,428 @@
+            let task_ptr = task.as_ref() as *const _;
+            unsafe {
+                fdb::fdb_future_set_callback(self.f, Some(fdb_future_callback), task_ptr as *mut _);
+            }
+            self.task = Some(task);
+
+            return Ok(Async::NotReady);
+        }
+
+        let ready = unsafe { fdb::fdb_future_is_ready(self.f) };
+        if ready == 0 {
+            return Ok(Async::NotReady);
+        }
+
+        let err = unsafe { fdb::fdb_future_get_error(self.f) };
+        if err != 0 {
+            return Err(FdbError::from(err));
+        }
+
+        let g = FdbFutureResult::new(self.f);
+        self.f = std::ptr::null_mut();
+
+        Ok(Async::Ready(g))
+    }
+}
+
+extern "C" fn fdb_future_callback(
+    _f: *mut fdb::FDBFuture,
+    callback_parameter: *mut ::std::os::raw::c_void,
+) {
+    let task: *const futures::task::Task = callback_parameter as *const _;
+    let task: &futures::task::Task = unsafe { std::mem::transmute(task) };
+    task.notify();
+}
+
+//TODO: impl Future
+fn example_set_get() -> Box<Future<Item = (), Error = FdbError>> {
+    let fut = FdbCluster::new("/etc/foundationdb/fdb.cluster")
+        .and_then(|cluster| cluster.create_database())
+        .and_then(|db| result(db.create_trx()))
+        .and_then(|trx| {
+            trx.set(b"hello", b"world");
+            trx.commit()
+        })
+        .and_then(|db| result(db.create_trx()))
+        .and_then(|trx| trx.get(b"hello"))
+        .and_then(|res| {
+            let val = res.value();
+            eprintln!("value: {:?}", val);
+
+            let trx = res.transaction();
+            trx.clear(b"hello");
+            trx.commit()
+        })
+        .and_then(|db| result(db.create_trx()))
+        .and_then(|trx| trx.get(b"hello"))
+        .and_then(|res| {
+            eprintln!("value: {:?}", res.value());
+            Ok(())
+        });
+
+    Box::new(fut)
+}
+
+fn example_get_multi() -> Box<Future<Item = (), Error = FdbError>> {
+    let fut = FdbCluster::new("/etc/foundationdb/fdb.cluster")
+        .and_then(|cluster| cluster.create_database())
+        .and_then(|db| result(db.create_trx()))
+        .and_then(|trx| {
+            let keys: &[&[u8]] = &[b"hello", b"world", b"foo", b"bar"];
+
+            let futs = keys.iter().map(|k| trx.get(k)).collect::<Vec<_>>();
+            join_all(futs)
+        })
+        .and_then(|results| {
+            for (i, res) in results.into_iter().enumerate() {
+                eprintln!("res[{}]: {:?}", i, res.value());
+            }
+            Ok(())
+        });
+
+    Box::new(fut)
+}
+
+fn main() {
+    let mut core = tokio_core::reactor::Core::new().unwrap();
+
+    let handle = unsafe {
+        let version = fdb::fdb_get_max_api_version();
+        let err = fdb::fdb_select_api_version_impl(version, version);
+        if err != 0 {
+            panic!("fdb_select_api_version: {:?}", FdbError::from(err));
+        }
+
+        let err = fdb::fdb_setup_network();
```

Comment:
Do you have any thoughts on #4 ? I want to try and make some safety guarantees around initialization. Everything in FoundationDB seems to be static, which is annoying.

Reply:
How about benchmarking `env_logger`? For me `env_logger` looks quite similar with `Network`.

 - Both `env_logger` / `Network` handles global static resources, and both could be initialized only once, with only one config.
 - `env_logger` should not be initialized by libraries, and I believe it also applies to `Network` too.

For now, I thinks it's better to expose those APIs with static functions, like `env_logger::{init/try_init}`.

```rust
use foundationdb as fdb;
fn main() {
  fdb::init();
  let handle = fdb::start_network();
  // do something
  fdb::stop_network();
  handle.join().expect("failed to join");
}
```

If user wants to customize a network config, they can use `Builder`, just like `env_logger::Builder`.

```rust
use foundationdb as fdb;
fn main() {
  let builder = fdb::Builder().set_network_version(500).set_option(fdb::options::NetworkOption::TraceEnable(true));
  builder.init(); // panic if called twice, just like `env_logger::Builder::init()`
  let handle = fdb::start_network();
  // do something
  fdb::stop_network();
  handle.join().expect("failed to join");
}
```

It seems that calling `fdb_run_network` multiple times causes segfault in library. I think we can fix the problem by having global state of whether the network is running or not.